lib/ocpp/v201/smart_charging.cpp

```diff
@@ -367,11 +369,19 @@ SetChargingProfileResponse SmartChargingHandler::add_profile(ChargingProfile& pr
                      [&profile](const ChargingProfile existing_profile) { return profile.id == existing_profile.id; });
 
     // K01.FR05 - replace non-ChargingStationExternalConstraints profiles if id exists.
-    if (profile_with_id != profile_storage.end() &&
-        profile_with_id->chargingProfilePurpose != ChargingProfilePurposeEnum::ChargingStationExternalConstraints) {
-        *profile_with_id = profile;
-    } else {
-        profile_storage.push_back(profile);
+    try {
+        // K01.FR27 - add profiles to database when valid
+        this->database_handler->insert_or_update_charging_profile(evse_id, profile);
+
+        if (profile_with_id != profile_storage.end() &&
+            profile_with_id->chargingProfilePurpose != ChargingProfilePurposeEnum::ChargingStationExternalConstraints) {
+            *profile_with_id = profile;
+        } else {
+            profile_storage.push_back(profile);
+        }
+
+    } catch (const QueryExecutionException& e) {
+        EVLOG_warning << "Could not store ChargingProfile in the database: " << e.what();
     }
```

Comment:
Let's remember to ask the Everest community about if it makes sense to only store and load profiles from the database instead of also keeping track within memory.

Reply:
It is noted. My plan is to ask if there is a reason for the duplication and depending on the answer probably move the conversation to a GitHub issue. It won't be a top priority for us at the moment.